lib/Stan_math_code_gen.ml

```diff
@@ -478,18 +453,63 @@ let trans_checks s =
   | _ -> s
 
 let%expect_test "trans check" =
-  let ccfunname = "greater_or_equal"
-  and ccargs = [zero]
-  and cctype = Ast.SInt in
-  let ck = {stmt= Check {ccfunname; ccvid= "N"; cctype; ccargs}; sloc= ""} in
-  print_s [%sexp ((trans_checks ck).stmt : stmt_loc statement)] ;
+  let ck =
+    Check
+      { ccfunname= "greater_or_equal"
+      ; ccvid= "N"
+      ; cctype= Ast.SInt
+      ; ccargs= [zero] }
+  in
+  print_s [%sexp ((trans_checks (with_no_loc ck)).stmt : stmt_loc statement)] ;
   [%expect
     {| (NRFunApp check_greater_or_equal ((Var function__) (Var N) (Lit Int 0))) |}]
 
-let add_data_read_mir s = match s.stmt with Decl _ -> s | _ -> s
+let decls_of_p {datavars; _} =
+  Map.Poly.data datavars |> List.map ~f:tvdecl_to_decl
+
+let pp_ctor ppf p =
+  pf ppf
+    {|
+%s(@[<v 0>stan::io::var_context& context__,
+@ unsigned int random_seed__ = 0;
+@ std::ostream* pstream__ = nullptr) : prob_grad(0) {@]
+@ @[<v 2>
+@ typedef double local_scalar_t__;
+@ boost::ecuyer1988 base_rng__ =
+@     stan::services::util::create_rng(random_seed__, 0);
+@ (void) base_rng__;  // suppress unused var warning
+@ static const char* function__ = "%s_model_namespace::%s";
+@ (void) function__;  // dummy to suppress unused var warning |}
+    p.prog_name p.prog_name p.prog_name ;
+  pf ppf
+    {|
+@ @ // Read in data variables
+@ %a
+@ @ // Transform data variables?
+@]
+|}
+    (list ~sep:cut pp_read_data)
+    (decls_of_p p)
+
+let pp_model_private ppf p =
+  pf ppf "@ %a" (list ~sep:cut pp_decl) (decls_of_p p)
+
+let pp_model_public ppf p = pf ppf "@ %a" pp_ctor p
 
 (* XXX *)
-let pp_model ppf p = ignore (ppf, p)
+let pp_model ppf p =
+  pf ppf
+    {|
+  class %s_model : public prob_grad {
+@ @[<v 1>
+@ private:
+@ @[<v 1>%a@]
```

Comment:
I'm going to need more time to figure all these out.  Was there doc somewhere for this fmt thing other than just the command-level doc?

Reply:
Well, it seems like `Fmt` mostly just changed the available functions but not the format specifiers/strings/whatever you call them. So `%a` is still that magic compositional format specifier from the [`Format`](https://caml.inria.fr/pub/docs/manual-ocaml/libref/Format.html) library. And `@` is a pretty-print specifier that works similarly to `%` for format strings - the first character that follows a `@` is always part of the pretty-printing command, most of which are documented on the ocaml Format module page linked above. There is a 1-to-1 correspondence between `@x` and the functions available in the Format library. On this line we have a 'space' break hint, equivalent to `print_space ()` followed by an `open_vbox 1`, followed by %a and the closing of the vbox. The complete correspondence is here:
![image](https://user-images.githubusercontent.com/805356/52544924-e656f380-2d81-11e9-9302-0f4aeaece474.png)


We could also switch to stop using all of the `@` pretty-printing specifiers, instead using explicit named functions instead (with a bit of an increase in verbosity). I had been thinking about that and don't feel strongly either way (maybe an experiment is warranted). 

Or just switch to lists of strings 😓